3단계 - 질문 삭제하기 리팩터링

victor-jo opened the pull request:
3단계 - 질문 삭제하기 리팩터링 리뷰 요청드립니다!

메시지 처리로 할 수 있는 부분들 리팩터링 해보았습니다!

잘부탁드립니다!

victor-jo committed:
refactor(answer, answers): 답변 관련 엔티티 리팩터링

- 답변 엔티티 삭제 후 히스토리 반환하는 메시지 추가
- 답변들 엔티티 전체가 주인인지 확인하는 메시지 추가
- 답변들 엔티티 전체 삭제 후 히스토리 반환하는 메시지 추가
- 답변들 엔티티 개수 반환하는 메시지 추가
- 질문 엔티티에서 답변 개수 가져오는 메시지 추가
- 질문 엔티티 테스트 리팩터링

victor-jo committed:
refactor(question): 질문 엔티티 리팩터링

- 질문 엔티티 삭제 후 히스토리 가져오는 메시지 추가
- 질문 엔티티에서 답변들이 모두 Owner 인지 확인하는 메시지 추가
- 질문 엔티티에서 답변 모두 삭제 후 히스토리 가져오는 메시지 추가

victor-jo committed:
refactor(qnaService): QnA 프로덕트 서비스 리팩터링

- 통일되게 질문 엔티티로 부터 메시지 발행시켜 처리하도록 수정

victor-jo committed:
refactor(qnaService, question): 질문 유효성검사 메시지 처리

- 질문 엔티티에서 유효성검사를 수행하도록 메시지로 변경

victor-jo committed:
refactor(qnaService, question): 질문 삭제 메시지 처리

- 질문 삭제 및 히스토리를 가져오는 메시지 통합

minzzang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -108,4 +110,9 @@ public String toString() {
     }
+
+    public DeleteHistory deleteAndGetHistory() {
+        setDeleted(true);
```
Answer에 있는 사용하지 않는 setter를 정리하고 필요한 부분은 의미있는 메소드로 작성해보면 어떨까요~?

victor-jo replied:
네!! 최대한 setter 메소드 정리하고 의미있는 이름으로 바꾸어 보았습니다~!

질문과 답변 엔티티 결과적으로 확인해보니 setter 메소드는 setContents 만 남은거 같아서 modify 로 변경하였습니다!

minzzang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -4,9 +4,11 @@
 import qna.UnAuthorizedException;
```
요구 사항 정리가 누락된 것 같네요.. 😄

victor-jo replied:
허억...! 넵 ㅠ 해당 부분 맨 처음부터 들어갔어야 하는데 추가하였습니다!

minzzang commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -35,22 +31,22 @@ public void refreshAnswerWithoutDelete() {
-    @Override
-    public Spliterator<Answer> spliterator() {
-        return Iterable.super.spliterator();
+    public List<DeleteHistory> allDeleteAndGetHistory() {
```
메소드명에 getHistory는 없어도 될 것 같은데 원준님은 어떻게 생각하시나요~?

victor-jo replied:
더 짧고 간결하고 좋은 것 같습니다~ 감사합니다~

minzzang commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -89,8 +89,8 @@ void setUp() {
     void 질문에서_참조를_통해_답변을_가져올때_삭제된_답변은_가져오지_않는다() {
         question.addAnswer(answer);
-        assertThat(question.getAnswers()).hasSize(1);
+        assertThat(question.getAnswersCount()).isEqualTo(1);
```
Answers에 대한 테스트가 없는 것 같네요! 그리고 삭제하기 기능에 대한 테스트가 전체적으로 없는 것 같습니다~

victor-jo replied:
넵 말씀해 주셔서 감사합니다

기능 요구사항에 맞춰 기능정의 설계한 그대로 테스트 케이스 추가하여 테스트하였습니다!

minzzang commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -33,24 +30,7 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        question.validateDelete(loginUser);
```
서비스 로직이 엄청 간결해졌네요! 👍

minzzang reviewed (changes requested):
안녕하세요 원준님~!

3단계 미션도 잘 작성해주셨네요! 👍 

소소한 코멘트 남겼는데 확인 부탁드립니다. :pray:

victor-jo committed:
refactor(question, answer): 객체 메소드 이름 변경

- 불필요해 보이는 메소드 이름을 간결하게 축약

before:
    deleteAndGetHistory 에서
after:
    delete 으로 모두 변경

victor-jo committed:
style(question): 불필요한 메소드 제거

victor-jo committed:
docs(README): 기능요구사항, 기능정의 정리

victor-jo committed:
test(answer): 답변 엔티티 메시지 추가 테스트

- 답변 작성자인지 아닌지 확인하는 isOwner 메시지 추가
- 답변을 삭제하는 delete 메시지 추가

victor-jo committed:
test(answer): 답변 엔티티 isOwner 메시지 리팩터링

- given 으로 전달되는 작성자를 저장 후 전달

victor-jo committed:
refactor(answers, question): 불필요한 메소드 제거

- 질문에서 답변을 삭제하는 없는 기능 제거
- 프로덕트 서비스에서 사용하지 않는 메소드 및 테스트 케이스 제거

victor-jo committed:
test(answers): 답변들 엔티티 테스트 케이스 추가

- 답변들의 작성자가 전달된 유저와 전체 일치하는지 확인하는 allOwner 메시지 추가
- 데이터를 전체 삭제하는 delete 메시지 추가

victor-jo committed:
refactor(question): 질문 엔티티 접근제한자 변경

victor-jo committed:
refactor(question): 질문 엔티티 테스트 케이스 리팩터링

- 의미를 부여할 수 있는 변수들은 의미 부여
- 잘못 테스트 되고 있는 로직 수정

victor-jo committed:
test(question): 질문 엔티티 테스트 케이스 추가

- 삭제 전 질문과 답변 작성자가 일치하는지 검증하는 validateDelete 메시지 추가
- 질문과 답변 데이터를 모두 삭제하는 delete 메시지 추가

victor-jo committed:
refactor(answer): 답변 엔티티 메시지 이름 변경

- 답변 내용을 수정하는 setContents 메시지를 modify 로 변경

before:
    setContents 에서
after:
    modify 으로 변경

victor-jo committed:
refactor(question): 질문 엔티티 메시지 이름 변경

- 질문 내용을 수정하는 setContents 메시지를 modify 로 변경

before:
    setContents 에서
after:
    modify 으로 변경

victor-jo commented:
말씀해 주신 부분들 수정하였습니다~ 바쁘신 와중에 피드백 하시느라 너무 ㅠ 고생하시는 것 같습니다

재리뷰 요청드리오니 잘 부탁드립니다!

minzzang commented on docs/README.md:
```diff
@@ -0,0 +1,19 @@
+  - 데이터를 삭제한다 (이력 정보 한개를 함께 준다)
+- 답변들 객체
+  - 답변들의 작성자가 전달된 유저와 전체 일치하는지 확인한다
+  - 데이터를 전체 삭제한다 (이력 정보를 리스트로 준다)
```
요구사항 정리 👍

minzzang commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,71 @@
+
+@DataJpaTest
+@Import(JpaAuditingConfiguration.class)
+public class AnswersTest {
```
test 추가 👍

minzzang commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Test
+    void 답변들을_모두_삭제한다() {
```
호불호가 갈리긴 하지만 given, when, then을 주석으로 달아 가독성을 높혀보는 방법을 사용해보셔도 좋을 것 같아요!
```
// when
List<DeleteHistory> deleteHistories = answers.delete();

// then
assertThat(deleteHistories).hasSize(4);
assertThat(deleteHistories).isEqualTo(Arrays.asList(
    new DeleteHistory(ANSWER, answerEntityList.get(0).getId(), writer, LocalDateTime.now()),
    new DeleteHistory(ANSWER, answerEntityList.get(1).getId(), writer, LocalDateTime.now()),
    new DeleteHistory(ANSWER, answerEntityList.get(2).getId(), writer, LocalDateTime.now()),
    new DeleteHistory(ANSWER, answerEntityList.get(3).getId(), writer, LocalDateTime.now())
));
```

minzzang reviewed (approved):
안녕하세요 원준님~!

피드백 잘 반영해주셨네요!! :+1:

jpa 미션이 짧은 미션이라 더 많은 대화를 나누지 못해 아쉽네요.. :pray:
이번 미션을 통해 jpa사용법과 테스트하기 쉬운 코드를 작성하는데 도움이 되셨으면 좋겠습니다!

저와의 미션은 여기서 끝이지만 다른 미션 진행하면서도 궁금한 점이나 어려운 부분 있으면
편하게 dm 해주세요 :smile:

jpa 미션 수행하시느라 정말 고생많으셨습니다.
남은 미션들도 모두 완주하셔서 좋은 결과있길 바라겠습니다! :bow:
